alf/algorithms/algorithm.py

```diff
@@ -1531,6 +1708,237 @@ def _update(self, experience, batch_info, weight):
+        else:
+            batch_indices = indices[mini_batch_start_position:min(
+                batch_size, mini_batch_start_position + mini_batch_size)]
+
+        def _make_time_major(nest):
+            """Put the time dim to axis=0."""
+            return alf.nest.map_structure(lambda x: x.transpose(0, 1), nest)
+
+        mini_batch_list = []
+        mini_batch_info_list = []
+        for experience, batch_info in zip(experience_list, batch_info_list):
+            if experience is not None:
+                batch = alf.nest.map_structure(lambda x: x[batch_indices],
+                                               experience)
+                if batch_info:
+                    binfo = alf.nest.map_structure(
+                        lambda x: x[batch_indices] if isinstance(
+                            x, torch.Tensor) else x, batch_info)
+                else:
+                    binfo = None
+                batch = _make_time_major(batch)
+                batch = alf.nest.utils.convert_device(batch)
+            else:
+                batch = None
+                binfo = None
+
+            mini_batch_list.append(batch)
+            mini_batch_info_list.append(binfo)
+
+        return mini_batch_list, batch_info_list
+
+    def _train_hybrid_experience(
+            self, experience, batch_info, offline_experience,
+            offline_batch_info, num_updates, mini_batch_size,
+            mini_batch_length, update_counter_every_mini_batch):
+        """Train using both experience (if available) and offline_experience.
+        We assume that experience can be None.
+        """
+        if experience is not None:
+            experience, processed_exp_spec, batch_info, length, mini_batch_length, batch_size = self._prepare_experience_data(
+                experience, self.experience_spec, batch_info,
+                mini_batch_length, self._replay_buffer)
+
+            self._processed_experience_spec = processed_exp_spec
+
+        # TODO: use a different mini_batch_length for offline training
+        offline_experience, _, offline_batch_info, length, mini_batch_length, batch_size = self._prepare_experience_data(
+            offline_experience, self._offline_experience_spec,
+            offline_batch_info, mini_batch_length, self._offline_replay_buffer)
+
+        indices = None
+        for u in range(num_updates):
+            if mini_batch_size < batch_size:
+                # here we use the cpu version of torch.randperm(n) to generate
+                # the permuted indices, as the cuda version of torch.randperm(n)
+                # seems to have a bug when n is a large number, generating
+                # negative or very large values that cause out of bound kernel
+                # error: https://github.com/pytorch/pytorch/issues/59756
+                indices = alf.nest.utils.convert_device(
+                    torch.randperm(batch_size, device='cpu'))
+
+            for b in range(0, batch_size, mini_batch_size):
```

Comment:
I am still not sure why we need to separate this training loop from the RL one

Reply:
Same as above, the main idea in design is to introduce minimal impacts on existing methods.